### PR TITLE
New metadata API: add MetaFile and TargetFile classes

### PR DESCRIPTION
**Pull request by MVrachev**
This pr is a continuation from pr https://github.com/theupdateframework/tuf/pull/1223

This is the final pr planned for https://github.com/theupdateframework/tuf/issues/1139, 

**Description of the changes being introduced by the pull request**:

In the top-level metadata classes, there are complex attributes such as
"meta" in Targets and Snapshot, "key" and "roles" in Root etc.
We want to represent those complex attributes with a class to allow
easier verification and support for metadata with unrecognized fields.
For more context read ADR 0004 and ADR 0008 in the docs/adr folder.

**Please verify and check that the pull request fulfills the following
requirements**:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

### CODE REVIEW

File: tuf/api/metadata.py
Comment (sechkova) [inline tuf/api/metadata.py]: Can you add a new line before the 'Attributes' section?

File: tuf/api/metadata.py
Comment (sechkova) [inline tuf/api/metadata.py]: After abstracting the (de)serialization, other formats than `json` are possible and although the default is `json` it shouldn't be hardcoded.

File: tuf/api/metadata.py
Comment (sechkova) [inline tuf/api/metadata.py]: 'meta' is already listed as class attribute above.

File: tuf/api/metadata.py
Comment (sechkova) [inline tuf/api/metadata.py]: Same comment as above, related to hardcoding file formats. In this case if an implementation does not use `.json` files this will probably raise a `KeyError`.
Possibly the filename/format should be read from `timestamp_dict` instead.

File: tuf/api/metadata.py
Comment (sechkova) [inline tuf/api/metadata.py:759]: This is not your change but again, hardcoded `.json`.

File: tuf/api/metadata.py
Comment (sechkova) [inline tuf/api/metadata.py]: The `__eq__` implementation is needed only for the tests, right?
I guess it doesn't hurt having it :shrug: Maybe it will come handy in future.

File: tuf/api/metadata.py
Comment (sechkova) [inline tuf/api/metadata.py]: Same comment about `json`.

**Review by sechkova — COMMENTED** [review]
In the PR description you say that it will:
> Rename a few instances of the update method in our classes and use a more verbose name.

but there is only one commit with marked TODOs.

My general comment is about removing hardcoded file extensions and mentions of "json" throughout the code. After the abstract (de)serialization was implemented, TUF supposedly supports other file formats too.

File: tuf/api/metadata.py
Comment (MVrachev) [reply tuf/api/metadata.py]: Yes, currently `__eq__` is needed only for tests.

File: tuf/api/metadata.py
Comment (MVrachev) [reply tuf/api/metadata.py:541]: That's a really good catch!
Will make it work with other extensions as well!

**Comment by MVrachev** [comment]
I rebased and fixed your two small comments @sechkova about other issues besides the hard-coded file names.

I fixed the problem of hardcoded `json` strings partially in the places I modified already.
To fully fix this problem we either have to use a more complicate solution or wait and consider other alternatives
which will give us the ability to set metadata extensions as a configuration option.

Created an issue about the hardcoded filename problems: https://github.com/theupdateframework/tuf/issues/1333

File: tuf/api/metadata.py
Comment (sechkova) [inline tuf/api/metadata.py]: This is a reasonable check but aren't we just assuming that input data is correct everywhere else? Since input validation is still a WIP? Anyway, probably better having it than not.

**Review by sechkova — APPROVED** [review]

File: tuf/api/metadata.py
Comment (MVrachev) [reply tuf/api/metadata.py]: Well the problem is that `meta_dict` sometimes is passed from the user - when calling `Timestamp.from_dict()`
and consequently when calling the `cls. _get_snapshot_filename()` method we haven't validated the `meta_dict` field.

At least, that is the case before implementing any new validation option as described in ADR 007.

File: tuf/api/metadata.py
Comment (jku) [inline tuf/api/metadata.py]: `Optional[Dict[str, str]]` I suppose

(much of the API uses Mapping in place of Dict though... I think that might be a mistake but maybe I've missed the reasons?)

File: tuf/api/metadata.py
Comment (jku) [inline tuf/api/metadata.py]: I think `get_snapshot_filename()` is not useful -- it does not do much and the ValueError is now hidden in an unexpected function. Overall we end up doing many (4?) dictionary lookups when I think one would be enough (it's not a performance issue or that important here but ... just more code than should be needed).

File: tuf/api/metadata.py
Comment (jku) [inline tuf/api/metadata.py]: We could just have `snapshot_filename` and `snapshot_info` attributes instead of the dictionary... 

That way the API would be more usable as well: `timestamp.signed.meta["snapshot.json"].version` is ugly and won't autocomplete whereas `timestamp.signed.snapshot_info.version` is ... well, harder to typo at least.

File: tuf/api/metadata.py
Comment (jku) [reply tuf/api/metadata.py]: Something like this (assuming the attribute changes I proposed above, and popping being acceptable):
```
meta_dict = timestamp_dict["meta"]
# TODO check length
snapshot_filename, snapshot_dict = meta_dict.popitem()
snapshot_info = MetadataInfo(**snapshot_dict)
return cls(*common_args, snapshot_filename, snapshot_info)
```

File: tuf/api/metadata.py
Comment (jku) [inline tuf/api/metadata.py]: `for meta_path, meta_dict in snapshot_dict["meta"].items()` -- no need to do the lookup later

File: tuf/api/metadata.py
Comment (jku) [inline tuf/api/metadata.py]: I don't know where this is used so this may be dumb, but is this function needed?
Would `timestamp.signed.snapshot_info = MetadataInfo(version, length, hashes)` be equivalent to to calling this ?

(this assumes the snapshot_info change proposed above)

File: tuf/api/metadata.py
Comment (jku) [inline tuf/api/metadata.py]: `for target_path, target_dict in targets_dict["targets"].items():`

File: tuf/api/metadata.py
Comment (jku) [inline tuf/api/metadata.py]: Same question here, is this a useful function vs caller modifying targets.signed.targets directly

File: tuf/api/metadata.py
Comment (jku) [inline tuf/api/metadata.py]: This works but is there a reason to use update over the simpler index access when you only have a single item?
```
 timestamp_dict["meta"] = {
     ...
}
```

**Review by jku — COMMENTED** [review]
Apologies for not looking at this before. I've gone through most, but maybe not quite processed all of it.

Martin knows of this but for others information: the changes to `.pop()` usage are going to clash with another PR #1345: they need to be coordinated. In general I think this PR could have been several smaller instead -- I get that it means annoying depending PRs... but this way means I feel bad for leaving so many comments :grin:  

Would be great to see a bit more prose about the design and usability of the specific API classes somewhere: If we go through the trouble of creating objects for everything, we should try to make them more usable than the dictionaries they replace. As examples, see question in code about snapshot_info attribute and ideas like _'should Targets maybe implement abc.Mapping so caller could access targets.signed["target_name"] instead of targets.signed.targets["target_name"]'_ -- that last one may not be a good idea but it would be good to see some ideas and discussion before we start maintaining more code.

This feels like a lot to ask on a big PR but would be nice to see how the code now copes without broken input in tests (missing fields, extra fields, etc) -- to confirm the benefits that we should be getting from this work.

**Comment by jku** [comment]
> Apologies for not looking at this before. I've gone through most, but maybe not quite processed all of it.

Also grumpy yesterday me sounds like he is unsatisfied with this PR but that's not true so I'll say it now:
PR looks correct to me (no bugs that I can find), good work, thank you

**Comment by jku** [comment]
> Would be great to see a bit more prose about the design and usability of the specific API classes somewhere: If we go through the trouble of creating objects for everything, we should try to make them more usable than the dictionaries they replace

After further discussion, this opinion seems to go against previous decisions to keep the API as close to file format as possible -- so possibly we need to discuss further or implement this as "faithful reproduction of spec" (disregarding my comments) and think about potential usability improvements afterwards...

File: tuf/api/metadata.py
Comment (jku) [reply tuf/api/metadata.py]: the above comment was done when I wasn't aware/thinking of the earlier decision to try to map the API to file format as close as possible... in that light you might want to disregard this (I think I still want the usable version but that can be added later as a property)

**Comment by MVrachev** [comment]
There were a lot of discussions after my last update on this pr.
I would prefer if we wait for #1345 and #1360 to be merged first.
Then, I will update/split this pr.

File: tuf/api/metadata.py
Comment (MVrachev) [reply tuf/api/metadata.py]: Agree, this could be subject to another set of changes.

File: tuf/api/metadata.py
Comment (MVrachev) [reply tuf/api/metadata.py]: I removed `get_snapshot_filename` from the latest rebased version.
I will think more about how to fix this approach when working on #1333

File: tuf/api/metadata.py
Comment (MVrachev) [reply tuf/api/metadata.py]: It will be the same I suppose, but I hope we will use this function to make some validation before doing the assignment.

File: tuf/api/metadata.py
Comment (MVrachev) [reply tuf/api/metadata.py]: It will be the same I suppose, but I hope we will use this function to make some validation before doing the assignment.

**Comment by MVrachev** [comment]
I have rebased and made changes in this pr after "develop" has evolved a lot.

Some of the changes I made:
- rename `MetadataInfo` to `MetaFile` and `TargetInfo` to `TargetFile`. Those names seem better in my opinion because they communicate better what the classes represent.
- added support for unrecognized fields in `MetaFile` and `TargetFile`. I didn't add a test, because #1382 will add an automatic test for that for dictionaries and classes.
- dropped a couple of the planned commits after #1367 was merged.
- dropped a couple more commits, because I realized they were not needed - removing the side effect of destroying the dictionaries when calling `from_dict()` from the classes (this is needed for ADR 8) and adding a todo comment about the `update()` functions naming (there is an issue for that #1230)
- added `from_dict()` for `MetaFile` and `TargetFile`
- added `MetaFile` and `TargetFile` specific tests
- disabled the `C0302: Too many lines in module` warning for tuf/api/metadata.py.

@jku and @sechkova pr is again ready for a fresh review.

File: tuf/api/metadata.py
Comment (jku) [inline tuf/api/metadata.py]: `for meta_path, metafile_dict in meta_dict.items():` or something like it maybe?

File: tuf/api/metadata.py
Comment (jku) [inline tuf/api/metadata.py]: nit: `snapshot_dict["meta"] = meta_dict` seems more normal

File: tuf/api/metadata.py
Comment (jku) [inline tuf/api/metadata.py]: `res_dict["targets"] = targets`

File: tuf/api/metadata.py
Comment (jku) [inline tuf/api/metadata.py]: This function should have TargetFile as argument, no?

File: tuf/api/metadata.py
Comment (jku) [inline tuf/api/metadata.py]: missing unrecognised_fields

File: tuf/api/metadata.py
Comment (jku) [inline tuf/api/metadata.py]: It could be argued we don't need to handle custom at all in any of these functions: it's just an unrecognised field and already supported via that mechanism.

If we want to match the file format exactly we'd then need a "custom" property getter though (that refers to `unrecognised_fields.get("custom")`)

No strong opinions, maybe I lean a bit towards the property :shrug:

File: tuf/api/metadata.py
Comment (jku) [inline tuf/api/metadata.py]: If we have this function it should take a Metafile as argument, IMO.

some other opinions on update():
* if we keep it a function, it should be renamed: "update" tells me nothing
* on the other hand I kind of think this should be a property instead, maybe "snapshot_meta" so you could do things like :
`version = timestamp.signed.snapshot_meta.version
`
or 
`timestamp.signed.snapshot_meta = MetaFile(ver, len, hashes)`

File: tuf/api/metadata.py
Comment (jku) [reply tuf/api/metadata.py]: I think a property checks all boxes: I'll leave a comment in the new code

File: tuf/api/metadata.py
Comment (jku) [reply tuf/api/metadata.py]: good point, thanks

File: tuf/api/metadata.py
Comment (jku) [reply tuf/api/metadata.py]: there may be an argument _against_ using helper properties like the one I handwaved: that it does not separate the attributes-that-map-the-file-format and the helpers enough (and that functions like `update()` are better in that regard). 

I think that's an issue we should consider but it does not outweigh the advantages (such as having a single read-and-write property with a descriptive name).

**Comment by jku** [comment]
A question on the `__eq__` implementations: do we need them? None of the other classes seem to have them -- I assumed you used them in the tests but that does not seem the case (or maybe I just missed it)

File: tuf/api/metadata.py
Comment (MVrachev) [reply tuf/api/metadata.py]: About the `update()` naming there is an issue already opened capturing this #1230.

I think we should restrain ourselves from changing the API here (like changing the argument to `MetaFile` or adding properties), but focus more on the changes included in the pr. 

Instead, those could be discussed in the issue I referenced above and the issue area of focus could widen to discuss the `update()` functions design as a whole.

File: tuf/api/metadata.py
Comment (MVrachev) [reply tuf/api/metadata.py]: I like the idea of using custom as `unrecognized fields` through using a property. 
Will do it and will add a comment that custom will replace `unrecognized fields` for this particular case.

Comment (MVrachev) [reply tuf/api/metadata.py]: Wondering, if we use `custom` replacing the need of `unrecognized fields` for this class, then do you test for this in #1382?
Do you add an unrecognized field under the `custom` dictionary?

File: tuf/api/metadata.py
Comment (MVrachev) [reply tuf/api/metadata.py]: I still think this is not in the scope of this pr.
See my comment https://github.com/theupdateframework/tuf/pull/1329#discussion_r631716043.

**Comment by MVrachev** [comment]
Updated the pr addressing almost all of the comments @jku made with the only exceptions when commenting on the `update` functions.
There I feel the changes Jussi proposed are out of scope for this pr as I described here https://github.com/theupdateframework/tuf/pull/1329#discussion_r631716043.

Additionally, I added a new commit where I remove the `__eq__` methods.

File: tuf/api/metadata.py
Comment (jku) [reply tuf/api/metadata.py:711]: I'm fine with not re-designing update here.

I disagree with leaving the argument as a dict. Why are we going through the trouble of adding objects and then not using them in the API?

File: tuf/api/metadata.py
Comment (jku) [reply tuf/api/metadata.py]: I'm fine with not completely re-designing update() methods in this PR.

I disagree with leaving the argument as a dict. Why are we going through the trouble of adding objects and then not using them in the API? 

I think your test code makes the case well: it does:
`targets.signed.update(filename, fileinfo.to_dict())`
and then in update you do 
` TargetFile.from_dict(fileinfo)`
it seems the argument should be TargetFile here

File: tuf/api/metadata.py
Comment (jku) [reply tuf/api/metadata.py]: Leaving update() re-design for later is fine.

I see no reason (so far) not to make the argument a MetaFile.

File: tuf/api/metadata.py
Comment (jku) [reply tuf/api/metadata.py]: I was trying to suggest dropping "custom", and just letting the already existing unrecognised_fields functionality handle it -- that basically handles de/serializing both "custom" and any other unrecognised fields.

Then I was saying that we might still want to add a property "custom" (to conform with the API design principle that the object API would copy the file format) but the property would just access `self.unrecognised_fields.get("custom")` and de/ser still would not handle "custom" at all -- it's handled like other unrecognised fields.

What you implemented is close but not quite the same... now TargetFile cannot have unrecognised fields at all (it can only have a "custom" field): as you correctly point out that means deciding that TargetFile is not allowed to have unrecognised fields and updating the test... which seems like the implementation detail guiding how we read the specification. 

I don't think this is a major issue, but I also think what I suggested would cover all the cases without any more complexity... opinions?

File: tuf/api/metadata.py
Comment (MVrachev) [reply tuf/api/metadata.py]: Okay, will change the argument type for the `update` functions, but won't rename it or add additional attributes as you suggested above.

File: tuf/api/metadata.py
Comment (MVrachev) [reply tuf/api/metadata.py]: I commented about the same question here:
https://github.com/theupdateframework/tuf/pull/1329#discussion_r632555437

File: tuf/api/metadata.py
Comment (MVrachev) [reply tuf/api/metadata.py]: Well, `custom` basically is the same as `unrecognised_fields`.
I feel like we should encourage our users to use `custom` as described in the spec when adding additional fields instead of relying on `unrecognized_fields` at least in this scenario.
Maybe we can achieve that by adding a `custom` property, but then if we want to treat it as a regular attribute maybe we would have to add a property setter as well?

On another hand, if we don't have `unrecognized_fields` in TargetFile, target.json file with unrecognized fields outside custom won't work...

Comment (jku) [reply tuf/api/metadata.py]: > Well, `custom` basically is the same as `unrecognised_fields`.
> I feel like we should encourage our users to use `custom` as described in the spec when adding additional fields instead of relying on `unrecognized_fields` at least in this scenario.
> Maybe we can achieve that by adding a `custom` property, but then if we want to treat it as a regular attribute maybe we would have to add a property setter as well?
> 
> On another hand, if we don't have `unrecognized_fields` in TargetFile, target.json file with unrecognized fields outside custom won't work...

that's what I'm saying: if we read the spec as "you need to be able to add new fields to any dictionary where it's technically possible" (which was the consensus I believe), then  implementing custom is not enough for spec-compliance

I don't think it's a major thing but ... if we can easily achieve spec compliance in a trivial matter then why not do it?

Comment (MVrachev) [reply tuf/api/metadata.py]: Okay, will update the class, so that `unrecognized fields` is documented and added as an attribute, but I am not sure should we document `custom` when `custom` will be just a property and we retrieve with `self.unrecognised_fields.get("custom")`?

Comment (MVrachev) [reply tuf/api/metadata.py]: I didn't define it in my last update. Do you think that's okay?

File: tuf/api/metadata.py
Comment (jku) [inline tuf/api/metadata.py]: please, no need to use update when a simple `timestamp_dict["meta"] = ...` works

File: tuf/api/metadata.py
Comment (jku) [inline tuf/api/metadata.py]: Should you have unrecognised_fields handling here instead of handling custom?

File: tests/test_api.py
Comment (jku) [inline tests/test_api.py]: This is not a dict like comment says but a MetaFile object, I think?

Is there a reason for this deepcopy and then modifying data -- couldn't you just create the object with MetaFile() using your arguments to make it clear what it is?

File: tests/test_api.py
Comment (jku) [inline tests/test_api.py]: Just a nit but mentioning in case there's something missing: This test doesn't really test much anything anymore (it used to compare two separate dicts, now it's pretty much comparing an object to itself)

File: tests/test_api.py
Comment (jku) [inline tests/test_api.py]: nit, This would be much more understandable as `fileinfo = MetaInfo(...)` as well

File: tests/test_api.py
Comment (jku) [inline tests/test_api.py]: same nit as before, this doesn't test much anything anymore

**Review by jku — CHANGES_REQUESTED** [review]
Everything else is a style comment (feel free to disregard) but I think there's a bug in the "custom" handling

File: tuf/api/metadata.py
Comment (MVrachev) [reply tuf/api/metadata.py]: I realized that I don't need to handle custom specifically anymore.
"custom", if exist, will be part of `unrecognized fields and my `to_dict` function could be just:
```
    def to_dict(self) -> Dict[str, Any]:
        """Returns the JSON-serializable dictionary representation of self."""
        return {
            "length": self.length,
            "hashes": self.hashes,
            **self.unrecognized_fields,
        }
```

File: tests/test_api.py
Comment (MVrachev) [reply tests/test_api.py]: Yes, I understand this. The test by itself makes sure that we are actually supporting optional `hashes` and `length`.
Will remove the test in the `fix update argument` commit. Then, we won't need this, because length and hashes will be optional because of the `MetaFile` implementation, no the update function itself.

**Comment by MVrachev** [comment]
Addressed all of @jku commets, and additionally I wanted to make sure that each of the commits passes the CI and makes sense.
That's why I also made the following changes:
1. Merge changes in "Replace usage of dict.copy with copy.deepcopy()" in "Add MetaFile" and "Add TargetFile". Then drop "Replace usage of dict.copy with copy.deepcopy()"
2. Simplified how we handle "custom" in `TargetFile.to_dict()`
3. Squash "New API: represent custom with unrecognized_fields" into "Add TargetFile"
4. Added a new commit removing two tests testing "update with only version" because they are obsolete.

File: tuf/api/metadata.py
Comment (jku) [reply tuf/api/metadata.py]: I see `__eq__()` is still here (which is fine -- maybe you use it in tests) but it's still not complete: it should compare unrecognized_fields instead of custom

**Review by jku — APPROVED** [review]
I think my comment in-source won't show up since it's in a resolved discussion but:

The` __eq__()` implementation is still not complete: if it is there, it should compare unrecognized_fields instead of custom.

Altogether this looks good, thanks. The docstrings are very verbose but let's trim those later... I'm marking this approved already, but please fix `__eq__()` still in this PR.

**Comment by MVrachev** [comment]
> I think my comment in-source won't show up since it's in a resolved discussion but:
> 
> The` __eq__()` implementation is still not complete: if it is there, it should compare unrecognized_fields instead of custom.
> 
> Altogether this looks good, thanks. The docstrings are very verbose but let's trim those later... I'm marking this approved already, but please fix `__eq__()` still in this PR.

I thought I had removed the `__eq__` methods.
It was a mistake that they are left there.
Removed them and rebased.

**Comment by jku** [comment]
I was about to merge but now it has conflicts with a bug fix I merged yesterday. Sorry. Could you rebase?

**Comment by MVrachev** [comment]
Done.

**Pull request merged**
